web/packages/teleterm/src/ui/ConnectMyComputer/connectMyComputerContext.tsx

```diff
@@ -97,69 +237,158 @@ export const useConnectMyComputerContext = () => {
 };
 
 /**
- * Waits for `error` and `exit` events from the agent process.
- * If none of them happen within 20 seconds, the promise resolves.
+ * Returns agent state based on multiple sources.
+ * Not all possible cases are handled, for example `download.success` -
+ * it is expected that the next state that the user should see is `start.processing`.
  */
-async function waitForAgentProcessErrors(
-  mainProcessClient: MainProcessClient,
-  rootClusterUri: RootClusterUri
-) {
-  let cleanupFn: () => void;
-
-  try {
-    const errorPromise = new Promise((_, reject) => {
-      const { cleanup } = mainProcessClient.subscribeToAgentUpdate(
-        rootClusterUri,
-        agentProcessState => {
-          const error = isProcessInErrorOrExitState(agentProcessState);
-          if (error) {
-            reject(error);
+function computeAgentState({
```

Comment:
My only gripe with `AgentState` is that I'm not sure if those specific enum values give us anything, that is we end up just repeating the error states of `Attempt`s.

If you look at `gzdunek/addcmc-status-document-wip`, my original idea was to return a step with the relevant attempt and other information needed to define the exact state of the agent.

```typescript
type Steps =
  | { step: 'download'; attempt: DownloadAttempt }
  | { step: 'join'; attempt: JoinAttempt };
```

Let's look at `NavigationMenu`:

https://github.com/gravitational/teleport/blob/cbbdeec78dd16f73fdd3d6c35b1e86365ac686e1/web/packages/teleterm/src/ui/ConnectMyComputer/NavigationMenu.tsx#L164-L172

This is the equivalent of having a case on `steps.step` and returning the blinking icon if the attempt associated with `download`, `start`, `kill` is in `processing` status.

`computeAgentState` has some nuances (like when `currentAttempt` is `start`) which, from what I can tell, are important only for the Status tab and mostly irrelevant for NavigationMenu and Setup. But since `computedAgentState` is shared across all three, it gives the impression that all three places care equally about those different states.

We can discuss this, but first I'd like to attempt to refactor this a little bit to put my money where my mouth is. ;f

My main concern is that whoever is going to have to maintain this feature will have to first understand everything I described in this comment before they unravel what AgentState is for. I think this could be simplified by not having separate enum states.

Reply:
>computeAgentState has some nuances (like when currentAttempt is start) which, from what I can tell, are important only for the Status tab and mostly irrelevant for NavigationMenu and Setup. But since computedAgentState is shared across all three, it gives the impression that all three places care equally about those different states.

Actually, this nuance is important for Setup too, it seems to me that it won't be easy to distinguish between 'join-error' / 'process-error' and 'process-exited' there, since the current attempt won't change from `start`?

> But since computedAgentState is shared across all three, it gives the impression that all three places care equally about those different states.

Yeah, they don't, but isn't it better to give the same, simple information to all three places and let them decide how they want to show it? It seems to me that after refactoring, we would have to return both the current step and all attempts so that callers have all the information they need to calculate the state. I am concerned that this places more responsibility on the `ConnectMyComputerContext` clients and would result in some duplication of logic.       

Also, I was experimenting with:
```
  const [currentStep, setCurrentStep] = useState<
    DownloadAttempt | JoinAttempt | ProcessAttempt
  >();
```
but I had a problem with keeping an `Attempt` object in the `currentStep` state - it was not reflecting changes that were done to the passed attempt. For example, when `downloadAttempt` changed from `''` to `processing`, the `currentStep.attempt` was still `''`.

--------
If you want to refactor the enum, I would start from handling the 'process-exited' error when the process fails to start.

Reply:
Did the refactor, let me know what you think. It's probably best to comment directly under the PR, or we can set up a call. https://github.com/gravitational/teleport/pull/30617